src/v/storage/segment_reader.cc

```diff
@@ -244,4 +244,86 @@ void segment_reader_handle::operator=(segment_reader_handle&& rhs) noexcept {
     _hook.swap_nodes(rhs._hook);
 }
 
+concat_segment_data_source_impl::concat_segment_data_source_impl(
+  std::vector<ss::lw_shared_ptr<segment>> segments,
+  size_t start_pos,
+  size_t end_pos,
+  ss::io_priority_class priority_class)
+  : _segments{std::move(segments)}
+  , _current_pos{_segments.begin()}
+  , _start_pos{start_pos}
+  , _end_pos{end_pos}
+  , _priority_class{priority_class} {}
+
+ss::future<ss::temporary_buffer<char>> concat_segment_data_source_impl::get() {
+    if (!_current_stream) {
+        _current_stream = co_await next_stream();
+    }
+
+    ss::temporary_buffer<char> buf = co_await _current_stream->read();
+    while (buf.empty() && _current_pos != _segments.end()) {
+        _current_stream = co_await next_stream();
+        buf = co_await _current_stream->read();
+    }
+
+    co_return buf;
+}
+
+ss::future<ss::input_stream<char>>
+concat_segment_data_source_impl::next_stream() {
+    if (_current_handle) {
+        co_await _current_handle->close();
+    }
+
+    vlog(
+      stlog.trace,
+      "opening stream for segment index {}, total segments: {}",
+      std::distance(_segments.begin(), _current_pos),
+      _segments.size());
+    auto segment = *_current_pos;
+    size_t start = 0;
+    size_t end = segment->file_size();
+
+    if (_current_pos == _segments.begin()) {
+        start = _start_pos;
+    }
+
+    if (std::distance(_current_pos, _segments.end()) == 1) {
+        end = _end_pos;
+    }
+
+    _current_pos++;
+
+    _current_handle = co_await segment->reader().data_stream(
+      start, end, _priority_class);
+    co_return _current_handle->take_stream();
+}
+
+ss::future<> concat_segment_data_source_impl::close() {
+    if (_current_handle) {
+        co_return co_await _current_handle->close();
+    }
+}
+
+concat_segment_reader_view::concat_segment_reader_view(
+  std::vector<ss::lw_shared_ptr<segment>> segments,
+  size_t start_pos,
+  size_t end_pos,
+  ss::io_priority_class priority_class)
+  : _stream(ss::data_source{std::make_unique<concat_segment_data_source_impl>(
+    std::move(segments), start_pos, end_pos, priority_class)}) {}
+
+ss::input_stream<char> concat_segment_reader_view::take_stream() {
+    auto r = std::move(_stream.value());
+    _stream.reset();
```

Comment:
Does resetting the optional here prevent closing of the last input stream? We call `close` after `take_stream`, so the `if (_stream)` check in `close` would never pass under normal operation and the last  input stream would be left open. Perhaps a destructor could be added `concat_segment_data_source_impl` to ensure that everything is closed.

Reply:
The way we use the handle interface right now in SI is that the stream is moved out of the handle in `remote.cc` and then that stream is closed separately, eg https://github.com/redpanda-data/redpanda/blob/ca7b635d50fef9c7f05b8ab6bbfe7255bc1f0598/src/v/s3/client.cc#L622 and the handle itself is also closed later (which is a no-op as the stream is moved out) https://github.com/redpanda-data/redpanda/blob/c2dcfcd8b5b129a06a973190fdf932251db8fe84/src/v/cloud_storage/remote.cc#L406


So the `close` on the view is just an extra check. If the stream has not been moved out it is closed in `concat_segment_reader_view::close()`.

Otherwise the consumer which moved the stream out of the view is responsible for closing it. Closing the stream closes the data source, which closes the data source impl `concat_segment_data_source_impl`, which in turn closes the last handle  `concat_segment_data_source_impl::_current_handle` if open. 

The current handle in the data source impl is closed when we switch to the next stream. I will add some comments to the implementation to explain this.

Reply:
Gotcha. Makes sense.